### ISSUE TITLE
scene_change.py optimizing tips?

### ISSUE THREAD

**Issue by Mr-Z-2697** [styler00dollar/VSGAN-tensorrt-docker#51]
Please forgive me for removing the non-onnx paths, and the reason I use DirectML is that it's probably the GPU provider easiest to setup on Windows.

I guess avoid the ndarray operations is the first key here, but I don't know if there's bad consequences to replace cv2.INTER_AREA (as it quite different to those resizing kernels).

Unfortunately
```
            I0 = np.expand_dims(I0, 0)
            I1 = np.expand_dims(I1, 0)
```
they sticks, as the model requires this I think unnecessary axis.

Input the pre-downscaled clip to the funtion also helps, or it probably can be done inside the function but at earlier stage.
(Input 1080p RGBS: ~90FPS, input 1080p YUV420P10: ~220FPS, input 256x256 RGBH: ~320FPS)

```
import numpy as np
import vapoursynth as vs
import functools
import onnxruntime as ort
from threading import Lock

def scene_detect(
    clip: vs.VideoNode,
    model_name: str = "efficientnetv2b0+rife46",
    thresh: float = 0.98,
    fp16: bool = True,
    onnx: bool = True,
    onnx_path: str = "test.onnx",
    onnx_res: int = 256,
    num_streams: int = 4,
    onnx_provider='Dml',
) -> vs.VideoNode:
    core = vs.core
    use_rife = False

    def frame_to_tensor(frame: vs.VideoFrame):
        return np.stack(
            [np.asarray(frame[plane]) for plane in range(frame.format.num_planes)]
        )

    def tensor_to_frame(f: vs.VideoFrame, array) -> vs.VideoFrame:
        for plane in range(f.format.num_planes):
            d = np.asarray(f[plane])
            np.copyto(d, array[plane, :, :])
        return f

    def tensor_to_clip(clip: vs.VideoNode, image) -> vs.VideoNode:
        clip = core.std.BlankClip(
            clip=clip, width=image.shape[-1], height=image.shape[-2]
        )
        return core.std.ModifyFrame(
            clip=clip,
            clips=clip,
            selector=lambda n, f: tensor_to_frame(f.copy(), image),
        )

    if not onnx:
        raise RuntimeError('non onnx removed')

    if onnx:
        sess = ort.InferenceSession(
            onnx_path,
            providers=[f"{onnx_provider}ExecutionProvider"],
        )
        resolution = onnx_res

    stream_lock = [Lock() for _ in range(num_streams)]
    index = -1
    index_lock = Lock()


    def execute(n: int, clip: vs.VideoNode) -> vs.VideoNode:
        nonlocal index
        with index_lock:
            index = (index + 1) % num_streams
            local_index = index

        if not onnx:
            raise RuntimeError('non onnx removed')
        if onnx:
            clip_down=clip.resize.Bicubic(resolution,resolution,format=vs.RGBH)
            I0 = frame_to_tensor(clip_down.get_frame(n))
            I1 = frame_to_tensor(clip_down.get_frame(n + 1))
            # I0 = np.rollaxis(I0, 0, 3)
            # I1 = np.rollaxis(I1, 0, 3)
            # I0 = I0.astype(np.float32)
            # I1 = I1.astype(np.float32)
            # I0 = cv2.resize(I0, (resolution, resolution), interpolation=cv2.INTER_AREA)
            # I1 = cv2.resize(I1, (resolution, resolution), interpolation=cv2.INTER_AREA)
            # I0 = np.expand_dims(I0, 0)
            # I0 = np.swapaxes(I0, 3, 1)
            # I0 = np.swapaxes(I0, 2, 3)

            # I1 = np.expand_dims(I1, 0)
            # I1 = np.swapaxes(I1, 3, 1)
            # I1 = np.swapaxes(I1, 2, 3)

            I0 = np.expand_dims(I0, 0)
            I1 = np.expand_dims(I1, 0)
            in_sess = np.concatenate([I0, I1], axis=1)
            result = sess.run(None, {"input": in_sess})[0]
            if result > thresh:
                return core.std.SetFrameProp(clip, prop="_SceneChangeNext", intval=1)
            else:
                return core.std.SetFrameProp(clip, prop="_SceneChangeNext", intval=0)
        return clip

    return core.std.FrameEval(
        core.std.BlankClip(clip=clip, width=clip.width, height=clip.height),
        functools.partial(execute, clip=clip),
    )
```
![2023-12-03_155817](https://github.com/styler00dollar/VSGAN-tensorrt-docker/assets/74594146/b4ed735d-4106-49bd-86ae-c41598d16727)
It was running like 18FPS here (also DirectML).

**Comment by Mr-Z-2697** [comment]
BTW, what's the difference with "+rife" variants (onnx ones specifically) and "_sim" variants?

**Comment by Mr-Z-2697** [comment]
If we move clip_down outside the FrameEval function
```
    clip_down=clip.resize.Bicubic(resolution,resolution,format=vs.RGBH)
    return core.std.FrameEval(
        core.std.BlankClip(clip=clip, width=clip.width, height=clip.height),
        functools.partial(execute, clip=clip, clip_down=clip_down),
    )
```
![2023-12-03_181257](https://github.com/styler00dollar/VSGAN-tensorrt-docker/assets/74594146/4b3a0f89-ed9f-40fe-8163-00f6977d0934)

**Comment by Mr-Z-2697** [comment]
If we use ModifyFrame instead of FrameEval...
```
    def execute(n,f):
        nonlocal index
        with index_lock:
            index = (index + 1) % num_streams
            local_index = index

        if not onnx:
            raise RuntimeError('non onnx removed')
        if onnx:
            fout=f[0].copy()
            I0 = frame_to_tensor(f[1])
            I1 = frame_to_tensor(f[2])
            I0 = np.expand_dims(I0, 0)
            I1 = np.expand_dims(I1, 0)
            in_sess = np.concatenate([I0, I1], axis=1)
            result = sess.run(None, {"input": in_sess})[0]
            if result > thresh:
                fout.props._SceneChangeNext=1
            else:
                fout.props._SceneChangeNext=0
        return fout
    clip_down=clip.resize.Bicubic(resolution,resolution,format=vs.RGBH)
    return core.std.ModifyFrame(clip,(clip,clip_down,clip_down[1:]),execute)
```
![2023-12-03_181640](https://github.com/styler00dollar/VSGAN-tensorrt-docker/assets/74594146/417e7894-3440-4a93-95cc-98b22640cb55)

**Comment by Mr-Z-2697** [comment]
BTW the get_frame in FrameEval function will cause the "Requesting frame number is beyond the last frame" error at last frame.

**Comment by Mr-Z-2697** [comment]
Actually, after moving resize out there's not much a difference between "original" input and 256x256 RGBH input now although the pre-downscaled and later prop-copied one is still slightly faster, but there should be some sort of matrix selecting (if not present in props).

**Comment by styler00dollar** [comment]
> don't know if there's bad consequences to replace cv2.INTER_AREA

Models were trained on area resize and I did not see such a scaling method with vapoursynth. That's why I tried to use area everywhere with OpenCV and torch. I did not test what will happen or if detection will be worse with other down scaling approaches. I guess a comparison is needed to make sure models properly work with vapoursynth resize or maybe [Kiyamou/VapourSynth-AreaResize](https://github.com/Kiyamou/VapourSynth-AreaResize) can be used.

> as the model requires this I think unnecessary axis.

Tensors require the shape `(batch, channel, height, width)`. It is required to do so and is standard practice with torch. I guess it would be possible to do `.unsqueeze(0)` in the torch code prior to exporting the onnx, since in this case batch size is always 1, but I am doubtful that this will make a significant change. Could be tested.

> what's the difference with "+rife" variants (onnx ones specifically) and "_sim" variants?

The rife variants additionally use the rife flow as input for the scene change classification model, so you don't only input images, but internally also input movement flow. `sim` refers to optimization with [daquexian/onnx-simplifier](https://github.com/daquexian/onnx-simplifier).

> If we use ModifyFrame instead of FrameEval...

In the past `FrameEval` was usually faster than `ModifyFrame` for me, that's why I didn't test it with the scene change code. It is interesting that `ModifyFrame` is faster here.

The approaches are quite interesting. I need more time to think about them and to test them though. Thanks for the feedback.

**Comment by Mr-Z-2697** [comment]
> Models were trained on area resize and I did not see such a scaling method with vapoursynth. That's why I tried to use area everywhere with OpenCV and torch. I did not test what will happen or if detection will be worse with other down scaling approaches. I guess a comparison is needed to make sure models properly work with vapoursynth resize or maybe Kiyamou/VapourSynth-AreaResize can be used.

I tested that plugin, its result is very similar to the opencv's, for fp32 only though (does not support fp16), other formats are quite different because of the gamma correction done by the plugin.

But with that the speed won't be as fast (~230FPS compared to ~600FPS). It's limited by CPU in this case, pushing my CPU to its limit while previously neither CPU nor GPU is heavily utilized. (resize.Bicubic to conver to RGBS, then area.AreaResize to resize, then inside the function the resize.Bicubic converts to RGBH)

**Comment by styler00dollar** [comment]
Did some testing with `core.num_threads = 4` on my 4090+7950x with `TensorrtExecutionProvider`, added clamp and softmax into onnx to have correct results.
```python
def execute(n: int, clip: vs.VideoNode) -> vs.VideoNode:
    clip_down=clip.resize.Bicubic(resolution,resolution,format=vs.RGBH)
    I0 = frame_to_tensor(clip_down.get_frame(n))
    I1 = frame_to_tensor(clip_down.get_frame(n + 1))
    in_sess = np.concatenate([I0, I1], axis=0)
    result = sess.run(None, {"input": in_sess})[0][0][0]
    if result > thresh:
        return core.std.SetFrameProp(clip, prop="_SceneChangeNext", intval=1)
    return clip

return core.std.FrameEval(
    core.std.BlankClip(clip=clip, width=clip.width, height=clip.height),
    functools.partial(execute, clip=clip),
)
```
```
sc_efficientnetv2b0_17957_256_CHW_6ch_clamp_softmax_op17_fp16_sim.onnx: 97.56 fps
sc_efficientformerv2_s0_12263_224_CHW_6ch_clamp_softmax_op17_fp16_sim.onnx: 92.87 fps
sc_swinv2_small_window16_10412_256_CHW_6ch_clamp_softmax_op17_fp16.onnx: 76.52 fps
```

```python
def execute(n: int, clip: vs.VideoNode, clip_down: vs.VideoNode) -> vs.VideoNode:
    I0 = frame_to_tensor(clip_down.get_frame(n))
    I1 = frame_to_tensor(clip_down.get_frame(n + 1))
    in_sess = np.concatenate([I0, I1], axis=0)
    result = sess.run(None, {"input": in_sess})[0][0][0]
    if result > thresh:
        return core.std.SetFrameProp(clip, prop="_SceneChangeNext", intval=1)
    return clip

clip_down=clip.resize.Bicubic(resolution,resolution,format=vs.RGBH)
return core.std.FrameEval(
    core.std.BlankClip(clip=clip, width=clip.width, height=clip.height),
    functools.partial(execute, clip=clip, clip_down=clip_down),
)
```
```
sc_efficientnetv2b0_17957_256_CHW_6ch_clamp_softmax_op17_fp16_sim.onnx: 97.23 fps
sc_efficientformerv2_s0_12263_224_CHW_6ch_clamp_softmax_op17_fp16_sim.onnx: 95.84 fps
sc_swinv2_small_window16_10412_256_CHW_6ch_clamp_softmax_op17_fp16.onnx: 81.75 fps
```

```python
def execute(n,f):
    fout=f[0].copy()
    I0 = frame_to_tensor(f[1])
    I1 = frame_to_tensor(f[2])
    I0 = np.expand_dims(I0, 0)
    I1 = np.expand_dims(I1, 0)
    in_sess = np.concatenate([I0, I1], axis=1)
    result = sess.run(None, {"input": in_sess})[0][0][0]
    if result > thresh:
        fout.props._SceneChangeNext=1
    return fout

clip_down=clip.resize.Bicubic(resolution,resolution,format=vs.RGBH)
return core.std.ModifyFrame(clip,(clip,clip_down,clip_down[1:]),execute)
```
```
sc_efficientnetv2b0_17957_256_BCHW_6ch_clamp_softmax_op17_fp16_sim.onnx: 289.56 fps
```

```python
def execute(n,f):
    fout=f[0].copy()
    I0 = frame_to_tensor(f[1])
    I1 = frame_to_tensor(f[2])
    in_sess = np.concatenate([I0, I1], axis=0)
    result = sess.run(None, {"input": in_sess})[0][0][0]
    if result > thresh:
        fout.props._SceneChangeNext=1
    return fout

clip_down=clip.resize.Bicubic(resolution,resolution,format=vs.RGBH)
return core.std.ModifyFrame(clip,(clip,clip_down,clip_down[1:]),execute)
```
```
sc_efficientnetv2b0_17957_256_CHW_6ch_clamp_softmax_op17_fp16_sim.onnx: 318.25 fps
sc_efficientformerv2_s0_12263_224_CHW_6ch_clamp_softmax_op17_fp16_sim.onnx: 320.63 fps
sc_swinv2_small_window16_10412_256_CHW_6ch_clamp_softmax_op17_fp16.onnx: 270.40 fps
```
ModifyFrame indeed seems a lot faster and removing that `expand_dims` also did give some improvement. Higher `num_threads` seems to improve results a bit on more, but I don't reach 600fps. It is still fast enough I guess, but I only have ~25% gpu utilization. For further speedup, using multiple sessions could be done similar to [this code](https://github.com/styler00dollar/vs-gmfss_fortuna/blob/a5325ab142a7a75106e05b37e2b4a04e75145ea8/vsgmfss_fortuna/__init__.py#L188) and [this code](https://github.com/styler00dollar/wd-v1-4-tagger-tensorrt-docker/blob/a295c77a3cd3b4bcfe2c5bc46fff04e4088ee785/tag.py#L103). Don't have time to test that for now.

Scene changes look good enough with bicubic. Can't get maxvit exported due to changes in timm. Will do the rife variants later. I uploaded the normal onnx for now and plan to update my code later. If I train such models again, I will do that with bicubic.

**Comment by Mr-Z-2697** [comment]
Thank you.

I still get 600 FPS with new onnx, if limit num_threads to 4 it drops to like 580 FPS. Perhaps DirectML is somehow more efficient than TenrorRT in this case?

**Comment by styler00dollar** [comment]
My benchmark code looks like this:
```python
import vapoursynth as vs
import functools
from src.scene_detect import scene_detect
core = vs.core
core.num_threads = 4 

clip = core.ffms2.Source(source="test.mkv", cache=False)
clip = vs.core.resize.Bicubic(clip, format=vs.RGBS, matrix_in_s="709")
clip = scene_detect(clip, thresh=0.95, resolution=256, onnx_path="sc_efficientnetv2b0_17957_256_CHW_6ch_clamp_softmax_op17_fp16_sim.onnx")
clip = vs.core.resize.Bicubic(clip, format=vs.YUV420P8, matrix_s="709")
clip.set_output()
```
```python
import numpy as np
import vapoursynth as vs
import onnxruntime as ort

def scene_detect(
    clip: vs.VideoNode,
    thresh: float = 0.98,
    onnx_path: str = "test.onnx",
    resolution: int = 256,
) -> vs.VideoNode:
    core = vs.core

    def frame_to_tensor(frame: vs.VideoFrame):
        return np.stack(
            [np.asarray(frame[plane]) for plane in range(frame.format.num_planes)]
        )

    sess = ort.InferenceSession(
        onnx_path,
        providers=['TensorrtExecutionProvider'] # CUDAExecutionProvider
    )  

    def execute(n,f):
        fout=f[0].copy()
        I0 = frame_to_tensor(f[1])
        I1 = frame_to_tensor(f[2])
        in_sess = np.concatenate([I0, I1], axis=0)
        result = sess.run(None, {"input": in_sess})[0][0][0]
        if result > thresh:
            fout.props._SceneChangeNext=1
        return fout

    clip_down=clip.resize.Bicubic(resolution,resolution,format=vs.RGBH)
    return core.std.ModifyFrame(clip,(clip,clip_down,clip_down[1:]),execute)
```
I currently can't test directml because I use linux.

I thought that adding YUV->RGBS->YUV would be some realistic workload since it most likely will get used with rife or something similar. If I remove that and add `matrix_in_s="709"`  into scene detect resize, I get:
```
sc_efficientnetv2b0_17957_256_CHW_6ch_clamp_softmax_op17_fp16_sim.onnx: 845.85 fps
sc_efficientformerv2_s0_12263_224_CHW_6ch_clamp_softmax_op17_fp16_sim.onnx: 859.76 fps
sc_swinv2_small_window16_10412_256_CHW_6ch_clamp_softmax_op17_fp16.onnx: 288.45 fps
``` 
My gpu usage is still 50-70% though.

**Comment by Mr-Z-2697** [comment]
Now that's some performance.

I use a Bicubic param like this btw, it kind of resembles the sharpness of area resize, maybe it would help, if that makes any sense.
https://github.com/Mr-Z-2697/z-vsPyScripts/blob/main/XcQ/scene_detect.py#L99C51-L99C90

**Comment by styler00dollar** [comment]
I tried to make it faster with multiple sessions.
```python
num_sessions = 3
sessions = [
    ort.InferenceSession(onnx_path, providers=["TensorrtExecutionProvider"])
    for _ in range(num_sessions)
]
sessions_lock = [Lock() for _ in range(num_sessions)]

index = -1
index_lock = Lock()

def execute(n, f):
    nonlocal index
    with index_lock:
        index = (index + 1) % num_sessions
        local_index = index

    fout = f[0].copy()
    I0 = frame_to_tensor(f[1])
    I1 = frame_to_tensor(f[2])
    in_sess = np.concatenate([I0, I1], axis=0)

    ort_session = sessions[local_index]
    result = ort_session.run(None, {"input": in_sess})[0][0][0]
    if result > thresh:
        fout.props._SceneChangeNext = 1
    return fout

clip_down = clip.resize.Bicubic(
    resolution, resolution, format=vs.RGBH, matrix_in_s="709"
)
return core.std.ModifyFrame(clip, (clip, clip_down, clip_down[1:]), execute)
```
```
sc_efficientnetv2b0_17957_256_CHW_6ch_clamp_softmax_op17_fp16_sim.onnx: 835.93 fps
sc_efficientformerv2_s0_12263_224_CHW_6ch_clamp_softmax_op17_fp16_sim.onnx: 935.89 fps
sc_swinv2_small_window16_10412_256_CHW_6ch_clamp_softmax_op17_fp16.onnx: 324.49 fps
```
`with sessions_lock[local_index]:` should be used, but I do not see performance gains with it. Having multiple sessions seems to give performance gains at least.

I don't think I can make it even faster and benchmarking gets quite unstable with such high numbers. Maybe int8 or sparsity can make further improvements, but I think this is good enough.

**Comment by styler00dollar** [comment]
Added optimized code into my repository and converted the other rife models. Closing issue.
